### DISCUSSION TITLE
Is there a way to figure out if an NDArray is trainable/has a gradient?

### DISCUSSION THREAD

**Discussion by theRoughCode** [apache/incubator-mxnet#19414] [Q&A] [answered]
Is there a way to figure out if an `NDArray` variable has a gradient/requires backprop, similar to the `requires_grad` attribute in PyTorch?

I've tried checking whether `x.grad` is `None`, but that doesn't work for intermediate variables:
```python
import mxnet as mx

x = mx.nd.array([1, 2])
x.attach_grad()
with mx.autograd.record():
  y = x**2
  z = 2 * y
z.backward()

>>> print(x.grad)
[4. 8.]
<NDArray 2 @cpu(0)>
>>> print(y.grad)
None
```

In this example, `x` has a gradient but the intermediate variable `y` does not, but it should also be trainable.

I've tried looking through the source code, but can't figure it out.

Corresponding SO post: https://stackoverflow.com/questions/64473958/is-there-a-way-to-figure-out-if-an-ndarray-is-trainable

**Comment by szha** [comment, marked as answer]
There isn't a direct way to access this information at the moment. It can be found in the backward operators as the req variable but it's not easy to tie this back to an NDArray.

What do you need this information for?

**Reply by theRoughCode** [reply]
Hey @szha, thanks for the response! I'm working on a Quantum Machine Learning library called PennyLane and we're planning on adding MXNet support as one of our ML backends. Specifically, we need to figure out which tensors are differentiable in order to perform automatic differentiation and avoid unnecessary computation for non-trainable tensors.

Are there any plans to add support for accessing this information (i.e. with an attribute similar to PyTorch)?
